File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/fencelits/domain/FenceLitsGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.fencelits.domain;

import com.rits.cloning.Cloner;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.Graph;
import com.zwstudio.logicpuzzlesandroid.common.domain.GridLineObject;
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions;
import com.zwstudio.logicpuzzlesandroid.common.domain.Node;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.F;
import fj.Ord;

import static fj.data.Array.array;
import static fj.data.HashMap.fromMap;
import static fj.data.List.iterableList;

/**
 * Created by zwvista on 2016/09/29.
 */

public class FenceLitsGameState extends CellsGameState<FenceLitsGame, FenceLitsGameMove, FenceLitsGameState> {
    public GridLineObject[][] objArray;
    public Map<Position, HintState> pos2state = new HashMap<>();

    public FenceLitsGameState(FenceLitsGame game) {
        super(game);
        objArray = new Cloner().deepClone(game.objArray);
        updateIsSolved();
    }

    public GridLineObject[] get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public GridLineObject[] get(Position p) {
        return get(p.row, p.col);
    }

    public boolean setObject(FenceLitsGameMove move) {
        Position p1 = move.p;
        int dir = move.dir, dir2 = (dir + 2) % 4;
        GridLineObject o = get(p1)[dir];
        if (o.equals(move.obj)) return false;
        Position p2 = p1.add(FenceLitsGame.offset[dir]);
        get(p2)[dir2] = get(p1)[dir] = move.obj;
        updateIsSolved();
        return true;
    }

    public boolean switchObject(FenceLitsGameMove move) {
        MarkerOptions markerOption = MarkerOptions.values()[game.gdi.getMarkerOption()];
        F<GridLineObject, GridLineObject> f = obj -> {
            switch (obj) {
            case Empty:
                return markerOption == MarkerOptions.MarkerFirst ?
                        GridLineObject.Marker : GridLineObject.Line;
            case Line:
                return markerOption == MarkerOptions.MarkerLast ?
                        GridLineObject.Marker : GridLineObject.Empty;
            case Marker:
                return markerOption == MarkerOptions.MarkerFirst ?
                        GridLineObject.Line : GridLineObject.Empty;
            }
            return obj;
        };
        GridLineObject[] dotObj = get(move.p);
        move.obj = f.f(dotObj[move.dir]);
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 12/FenceLits

        Summary
        Fencing Tetris

        Description
        1. The goal is to divide the board into Tetris pieces, including the
           square one (differently from LITS).
        2. The number in a cell tells you how many of the sides are marked
           (like slitherlink).
        3. Please consider that the outside border of the board as marked.
    */
    private void updateIsSolved() {
        isSolved = true;
        for (Map.Entry<Position, Integer> entry : game.pos2hint.entrySet()) {
            Position p = entry.getKey();
            int n2 = entry.getValue();
            int n1 = 0;
            for (int i = 0; i < 4; i++)
                if (get(p.add(FenceLitsGame.offset2[i]))[FenceLitsGame.dirs[i]] == GridLineObject.Line)
                    n1++;
            HintState s = n1 < n2 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            pos2state.put(p, s);
            if (n1 != n2) isSolved = false;
        }
        Graph g = new Graph();
        Map<Position, Node> pos2node = new HashMap<>();
        for (int r = 0; r < rows() - 1; r++)
            for (int c = 0; c < cols() - 1; c++) {
                Position p = new Position(r, c);
                Node node = new Node(p.toString());
                g.addNode(node);
                pos2node.put(p, node);
            }
        for (int r = 0; r < rows() - 1; r++)
            for (int c = 0; c < cols() - 1; c++) {
                Position p = new Position(r, c);
                for (int i = 0; i < 4; i++)
                    if (get(p.add(FenceLitsGame.offset2[i]))[FenceLitsGame.dirs[i]] != GridLineObject.Line)
                        g.connectNode(pos2node.get(p), pos2node.get(p.add(FenceLitsGame.offset[i])));
            }
        if (!isSolved) return;
        List<List<Integer>> fencelitses = new ArrayList<>();
        while (!pos2node.isEmpty()) {
            g.setRootNode(fromMap(pos2node).values().head());
            List<Node> nodeList = g.bfs();
            List<Position> area = fromMap(pos2node).toStream().filter(e -> nodeList.contains(e._2())).map(e -> e._1()).toJavaList();
            if (area.size() != 4) {isSolved = false; return;}
            area.sort(Position::compareTo);
            List<Position> treeOffsets = new ArrayList<>();
            Position p2 = new Position(iterableList(area).map(p -> p.row).minimum(Ord.intOrd),
                    iterableList(area).map(p -> p.col).minimum(Ord.intOrd));
            for (Position p : area)
                treeOffsets.add(p.subtract(p2));
            if (!array(FenceLitsGame.tetrominoes).exists(arr -> Arrays.equals(arr, treeOffsets.toArray()))) {
                isSolved = false; return;
            }
            for (Position p : area)
                pos2node.remove(p);
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/fencelits/domain/FenceLitsGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.fencelits.domain

import com.rits.cloning.Cloner
import com.zwstudio.logicpuzzlesandroid.common.domain.*
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState

class FenceLitsGameState(game: FenceLitsGame) : CellsGameState<FenceLitsGame, FenceLitsGameMove, FenceLitsGameState>(game) {
    var objArray: Array<Array<GridLineObject>> = Cloner().deepClone(game.objArray)
    var pos2state = mutableMapOf<Position, HintState>()

    operator fun get(row: Int, col: Int) = objArray[row * cols + col]
    operator fun get(p: Position) = this[p.row, p.col]

    init {
        updateIsSolved()
    }

    fun setObject(move: FenceLitsGameMove): Boolean {
        val dir = move.dir
        val dir2 = (dir + 2) % 4
        val p1 = move.p
        val p2 = p1.add(FenceLitsGame.offset[dir])
        val o = this[p1][dir]
        if (o == move.obj) return false
        this[p1][dir] = move.obj
        this[p2][dir2] = this[p1][dir]
        updateIsSolved()
        return true
    }

    fun switchObject(move: FenceLitsGameMove): Boolean {
        val markerOption = MarkerOptions.values()[game.gdi.markerOption]
        val o = this[move.p][move.dir]
        move.obj = when (o) {
            GridLineObject.Empty -> if (markerOption == MarkerOptions.MarkerFirst) GridLineObject.Marker else GridLineObject.Line
            GridLineObject.Line -> if (markerOption == MarkerOptions.MarkerLast) GridLineObject.Marker else GridLineObject.Empty
            GridLineObject.Marker -> if (markerOption == MarkerOptions.MarkerFirst) GridLineObject.Line else GridLineObject.Empty
            else -> o
        }
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 12/FenceLits

        Summary
        Fencing Tetris

        Description
        1. The goal is to divide the board into Tetris pieces, including the
           square one (differently from LITS).
        2. The number in a cell tells you how many of the sides are marked
           (like slitherlink).
        3. Please consider that the outside border of the board as marked.
    */
    private fun updateIsSolved() {
        isSolved = true
        // 2. The number in a cell tells you how many of the sides are marked
        // (like slitherlink).
        for ((p, n2) in game.pos2hint) {
            var n1 = 0
            for (i in 0 until 4)
                if (this[p.add(FenceLitsGame.offset2[i])][FenceLitsGame.dirs[i]] == GridLineObject.Line)
                    n1++
            val s = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            pos2state[p] = s
            if (n1 != n2) isSolved = false
        }
        val g = Graph()
        val pos2node = mutableMapOf<Position, Node>()
        for (r in 0 until rows - 1)
            for (c in 0 until cols - 1) {
                val p = Position(r, c)
                val node = Node(p.toString())
                g.addNode(node)
                pos2node[p] = node
            }
        for (r in 0 until rows - 1)
            for (c in 0 until cols - 1) {
                val p = Position(r, c)
                for (i in 0 until 4)
                    if (this[p.add(FenceLitsGame.offset2[i])][FenceLitsGame.dirs[i]] != GridLineObject.Line)
                        g.connectNode(pos2node[p]!!, pos2node[p.add(FenceLitsGame.offset[i])]!!)
            }
        if (!isSolved) return
        // 1. The goal is to divide the board into Tetris pieces, including the
        // square one (differently from LITS).
        while (pos2node.isNotEmpty()) {
            g.rootNode = pos2node.values.first()
            val nodeList = g.bfs()
            val area = pos2node.filter { nodeList.contains(it.value) }.map { it.key }.sorted()
            if (area.size != 4) {
                isSolved = false
                return
            }
            val treeOffsets = mutableListOf<Position>()
            val p2 = Position(area.map { it.row }.min()!!, area.map { it.col }.min()!!)
            for (p in area)
                treeOffsets.add(p.subtract(p2))
            if (!FenceLitsGame.tetrominoes.any { it == treeOffsets }) {
                isSolved = false
                return
            }
            for (p in area)
                pos2node.remove(p)
        }
    }
}